runtime/src/accounts.rs

```diff
@@ -320,36 +332,88 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, account);
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            let credit_only_locks = credit_only_locks.read().unwrap();
+            if locks.contains(k)
+                || (credit_only_locks.contains_key(k)
+                    && *credit_only_locks
+                        .get(&k)
+                        .unwrap()
+                        .lock_count
+                        .lock()
+                        .unwrap()
+                        > 0)
+            {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
                 error_counters.account_in_use += 1;
                 debug!("Account in use: {:?}", k);
                 return Err(TransactionError::AccountInUse);
             }
         }
-        for k in keys {
-            locks.insert(**k);
+
+        for k in credit_debit_keys {
+            locks.insert(*k);
+        }
+        let mut credit_only_writes: Vec<&Pubkey> = vec![];
+        for k in credit_only_keys {
+            let credit_only_locks = credit_only_locks.read().unwrap();
+            if let Some(credit_only_lock) = credit_only_locks.get(&k) {
+                *credit_only_lock.lock_count.lock().unwrap() += 1;
+            } else {
+                credit_only_writes.push(k);
+            }
+        }
+
+        for k in credit_only_writes.iter() {
+            let mut credit_only_locks = credit_only_locks.write().unwrap();
+            credit_only_locks
+                .entry(**k)
+                .or_insert_with(|| CreditOnlyLock {
+                    credits: AtomicU64::new(0),
+                    lock_count: Mutex::new(1),
+                });
```

Comment:
aren't you supposed to increment the lock_count if you find the entry, i.e. if another thread has beat you to building credit_only_writes?

Reply:
Yes, the increment happens above. There shouldn't be any way for the entry to be created by another thread between that point and this (see comment above).

Reply:
could put in an assert!() to that effect?

Reply:
Good idea